src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -229,42 +249,65 @@ protected int doWork() {
                 // First bring the simple metrics up to date
                 if (rec.getReadUnmappedFlag()) {
                     ++metrics.UNMAPPED_READS;
+                } else if(rec.isSecondaryOrSupplementary()) {
+                    ++metrics.SECONDARY_OR_SUPPLEMENTARY_READS;
                 } else if (!rec.getReadPairedFlag() || rec.getMateUnmappedFlag()) {
                     ++metrics.UNPAIRED_READS_EXAMINED;
                 } else {
                     ++metrics.READ_PAIRS_EXAMINED; // will need to be divided by 2 at the end
                 }
 
+            // Now try and figure out the next duplicate index (if going by coordinate. if going by query name, only do this
+            // if the query name has changed.
+            final boolean needNextDuplicateIndex = recordInFileIndex > nextDuplicateIndex &&
+                    (sortOrder == SAMFileHeader.SortOrder.coordinate || !rec.getReadName().equals(duplicateQueryName));
 
-                if (recordInFileIndex == nextDuplicateIndex) {
-                    rec.setDuplicateReadFlag(true);
+            if (needNextDuplicateIndex) {
+                    nextDuplicateIndex = (this.duplicateIndexes.hasNext() ? this.duplicateIndexes.next() : NO_SUCH_INDEX);
+            }
 
-                    // Update the duplication metrics
-                    if (!rec.getReadPairedFlag() || rec.getMateUnmappedFlag()) {
-                        ++metrics.UNPAIRED_READ_DUPLICATES;
-                    } else {
-                        ++metrics.READ_PAIR_DUPLICATES;// will need to be divided by 2 at the end
-                    }
+            final boolean isDuplicate = recordInFileIndex == nextDuplicateIndex ||
+                    (sortOrder == SAMFileHeader.SortOrder.queryname &&
+                    recordInFileIndex > nextDuplicateIndex && rec.getReadName().equals(duplicateQueryName));
 
-                    // Now try and figure out the next duplicate index
-                    if (this.duplicateIndexes.hasNext()) {
-                        nextDuplicateIndex = this.duplicateIndexes.next();
-                    } else {
-                        // Only happens once we've marked all the duplicates
-                        nextDuplicateIndex = -1;
+
+                if (isDuplicate) {
+                    duplicateQueryName = rec.getReadName();
+                    rec.setDuplicateReadFlag(true);
+
+                    // only update duplicate counts for "decider" reads, not tag-a-long reads
+                    if (!rec.isSecondaryOrSupplementary() && !rec.getReadUnmappedFlag()) {
```

Comment:
But for queryname sorted runs, 'tagalong' friends are being marked as duplicates right? So why wouldn't we want to update the count to include them? By doing this, will the duplicate metrics not match the count of duplicate tags derived from the file?